Performance-Tracing.md

```diff
@@ -0,0 +1,156 @@
+    7.  This metadata is specific to the kind of box you're looking at.
+        *This is the most important information* - it will contain paths and things.
+
+    *Note:* All times reflect the overhead of tracing, which is not spread uniformly through the execution.
+
+8.  As mentioned above, the checking phase will generally be the most interesting when diagnosing delays.
+    Here we see a zoomed in portion of the trace:\
+    ![Checking phase portion of trace](https://raw.githubusercontent.com/wiki/Microsoft/TypeScript/images/tracingCheckVariableDeclaration.png)
+    1.  What we're looking at is a call stack summary, with callers above and callees below.
+        *Not all frames are present* - the trace would be too large and hard to read.
+        Three box types are of particular interest when reducing a repro:
+        1.  `checkSourceFile` gives the time spent checking a particular file.
+            Look in the metadata for the file path.
+        2.  `checkExpression` gives the time spent checking a particular expression.
+            It will *usually, but not always* be in the file indicated by the containing `checkSourceFile`.
+            As of now, there is no way to tell without looking at the code and no way to recover the path if it's not the one from `checkSourceFile`.
+            We're thinking about how to handle it.
+        3.  `checkVariableDeclaration` gives the time spent checking a particular variable declaration.
+            Over the course of several investigations, we noticed that these were showing up often enough to merit their own boxes.
+            The metadata for these boxes is not particularly convenient to consume.
+            1.  `kind` identifies a member of the [SyntaxKind](https://github.com/microsoft/TypeScript/blob/9c60d5a4d33e4ed846cad2136da9efcd48adf847/src/compiler/types.ts#L21) enum.
+            2.  `pos` is the zero-indexed character offset from the beginning of the file to the beginning of the range being checked.
+                Note that position will be *before* any trivia (whitespace, comments) before the relevant code fragment, so it will frequently appears to be immediately after the *preceding* statement.
+            3.  `end` is the zero-indexed character offset from the beginning of the file to the end of the range being checked.
+    2.  In this particular screenshot, we see that a single variable declaration accounts for a large percentage of the check time of this entire file.
+        However, the declaration itself is not interesting because its child is an expression that is nearly as wide (i.e. the declaration has low "Self Time").
+        This expression has four interesting sub-expressions, each of which has many, many children.
+        So identifying those four ranges in the file would be the place to start when reducing a repro.
+
+9.  Once you have identified the problematic code, review [[Performance]] for suggestions relevant to your scenario.
```

Comment:
This link isn't working in the preview and I don't know if that's a preview issue or if the syntax is wrong.

Reply:
It works on the real page.